Added "First Command" guide to D.JS tutorial series.

NorteX-dev opened the pull request:
Hi.
I've been making a tutorial for making a very simple command in Discord.js.
Took me about 3-4 hours in total.
Covers adding a message event, splitting the content to command/args, and creating a simple if statement. Very in-depth, at least I believe.

NorteX-dev committed:
Update tableOfContents.md

NorteX-dev committed:
Create firstCommand.md

Tried my best on this, I hope it's fine. If it gets accepted I'll write more, if not then it's fine and I won't write anymore.

NorteX-dev committed:
Update firstCommand.md

NorteX-dev committed:
Update firstCommand.md

NorteX-dev committed:
Update Sidebar.js

NorteX-dev committed:
Update firstCommand.md

JonasSchiott reviewed (changes requested):
Message listener from line 80-82 will spam the chat as bot would respond to own message.

NorteX-dev committed:
Update firstCommand.md

Fixed so the bot doesn't fall into a self-responding loop.

JonasSchiott committed:
"Improved" if-statement

JonasSchiott commented:
@acollierr17 thoughts please

acollierr17 commented on guide/topics/discordjs/firstCommand.md:
````diff
@@ -0,0 +1,163 @@
+});
+```
+
+To make sure that the bot will not fall in a loop responding to it's own messages, we have to add an `if` statement check, to check if the author is a bot. If so, `return` the code. The same goes for the second line, which check's for the guild of the message, as we don't want to listen to DM messages. Now, that we are sure that the bot will not fall into a loop and that guild is present on the o ...
````
Here you state to check the second line. However, there's only one line, but two *conditions*. It should state to check the second *condition* which is `!message.guild`. This should help make the sentence more clear.

acollierr17 replied:
Also, since you're discussing the `.send()` method here, and this is in the context of a "textable" channel, I would link the documentation for [`TextChannel#send()`](https://discord.js.org/#/docs/main/stable/class/TextChannel?scrollTo=send) for example so readers can see various examples of using the method.

booleans-oss replied:
I don't think that the condition ``!message.guild`` really help not to get into a loop. Because, this condition only checks if the message has been posted on a guild which cannot really cause a loop. Also, if we keep this condition, we would need to say that it would ignore messages sent in the bot's DM.

acollierr17 replied:
> I don't think that the condition `!message.guild` really help not to get into a loop. Because, this condition only checks if the message has been posted on a guild which cannot really cause a loop. Also, if we keep this condition, we would need to say that it would ignore messages sent in the bot's DM.

That's a fair statement to make. In this case, you could put both conditions in their own `if` statements? Then you can explain the purpose of both conditions separately since in this tutorial, it only really matters if the message author is a bot or not.

acollierr17 commented on guide/topics/discordjs/firstCommand.md:
````diff
@@ -0,0 +1,163 @@
+
+Just to remind us, this is the code we had set up in the end:
+```javascript
+// Import the "Client" class from discord.js library
````
I would consider adding a hyperlink to the [`Client`](https://discord.js.org/#/docs/main/stable/class/Client) class as a quick reference.

acollierr17 commented on guide/topics/discordjs/firstCommand.md:
````diff
@@ -0,0 +1,163 @@
+// Authenticate the bot with your token from the Discord Developer Page
+client.login("-- Your Token --");
+```
+Upon running above code, the bot printed "I am ready!" in the console. Great!
````
I think this comment could be a bit more clear:
> Listen to the "ready" event and print "I am ready!" whenever the bot is connected to the Discord API

acollierr17 commented on guide/topics/discordjs/firstCommand.md:
````diff
@@ -0,0 +1,163 @@
+  console.log("Hello!");
+});
+```
+Let's brake this code down. As with the `ready` event, we are accessing the `client` variable. Then, we are using `.on()` which is a method that creates an _event listener_ on our Discord bot client.
````
- Type for the word `brake`. It should be `break`.

- Also since you mentioned message listeners/emitters, I would provide a hyperlink to the Node.js documentation on [`Events`](https://nodejs.org/api/events.html). Inline references like this can be really good for beginners to keep track of any new terminology they might pick up.

acollierr17 commented on guide/topics/discordjs/firstCommand.md:
```diff
@@ -0,0 +1,163 @@
+- Basics
+---
+
+Disclaimer: this tutorial assumes you've followed the [Getting Started](gettingStarted.md) tutorial.
```
Based on checking all the Discord.js documentation links, I would include an additional disclaimer that this tutorial is based on the `stable` branch of Discord.js. It's not the biggest deal, but some people might want to know that knowledge before reading and if they need to double-check with the `master` branch before proceeding.

acollierr17 commented on guide/topics/discordjs/firstCommand.md:
```diff
@@ -0,0 +1,163 @@
+
+The event name is obviously the name of the event that we want to listen to. There are many events, e.g. these which we have learnt already: `ready` and `message`, but also `guildCreate`, `guildMemberAdd`, `messageDelete`, `roleCreate` and many, many more. You can see all of them in this [Discord events cheatsheet by koad](https://gist.github.com/koad/316b265a91d933fd1b62dddfcc3ff584).
+
+The other parameter, a function, is the function that will be invoked when the event is fired. Oftentimes, it takes in parameters specific for that type of event. Our `message` event, only returns 1 parameter: a message object which we will be able to utilize in a second.
```
I recommend better wording here:
> a message object, which we will be able to utilize later in this guide.

Vexxlol replied:
> The other parameter, a function, is the function that will be....

That's a bit hard to read IMO. I suggest being more straight forward.
"The second parameter is a call-back. This will be executed once the event is triggered."

acollierr17 commented on guide/topics/discordjs/firstCommand.md:
```diff
@@ -0,0 +1,163 @@
+client.on("message", (message) => {
+  if(message.author.bot || !message.guild) return;
+  
+  const split = message.content.split(" ");
+  const command = split[0];
+  const args = split.slice(1);
```
If you want to refactor this into one line, you could do this:
```js
const [command, ...args] = message.content.split(" ");
```

The example above declares variables from `message.content.split(" ")` via "destructuring assignment". Each variable represents an index in the array. So `const command = split(0);` is the same as the `command` variable destructured above.

For more information on destructuring assignment, check out the [`MDN Docs`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment).

JonasSchiott replied:
Just commit it yourself, if it hasn’t been added yet.

acollierr17 commented on guide/topics/discordjs/firstCommand.md:
````diff
@@ -0,0 +1,163 @@
+});
+```
+
+Now, a very straightforward `if` statement to check if the `command` part is equal to "!ping". Notice how when checking, after both `command` and the literal command string, and sending a response. Very easy.
````
I think the paragraph's wording can be more straightforward:
> Now, a very straightforward `if` statement checks if the `command` string is equal to `!ping`. We make this comparison before sending the response to ensure the right command is ran. We also force it to lowercase to allow running commands without worrying about case sensitivity.

acollierr17 commented on guide/topics/discordjs/firstCommand.md:
````diff
@@ -0,0 +1,163 @@
+client.login("-- Your Token --");
+```
+
+In the next tutorial we'll learn how to learn [todo next tutorial]
+<!-- TODO : Add a link -->
````
Personally, I would remove this for now until there's a solid idea of what additional entries are planned after this one. I feel for these guides, consistency between different entries is paramount, so that's the main factor in my response here.

NorteX-dev committed:
Fixed readibility

- Split if statements into 2
- Added hyperlink to discord docs on send() method
- Added hyperlink to the Client class
- Replaced line 29 to be more clear
- break instead of brake
- Added hyperlink to Node's Events class
- Added disclaimer about stable branch
- Wording on lines 53/55 corrected
- Refactored code in lines >96
- Line 131 changed as @acollierr17 mentioned
- Removed footer until other tutorials are added
- Edited line 55 wording as @Vexxlol suggested

NorteX-dev commented:
Sorry for the delay, fixed all issues now (hopefully).

JonasSchiott committed:
Update firstCommand.md

Strandxo committed:
missing - causing error

Strandxo reviewed (changes requested):
Missing - on the Javascript category name... This is causing an issue with building the file(s)

Strandxo commented:
Altered it myself, simple fix. If changes are approved then Im going to merge. Thank you for your contribution.
